archive/tar_opts_windows.go

```diff
@@ -27,3 +27,22 @@ func AsWindowsContainerLayer() ApplyOpt {
 		return nil
 	}
 }
+
+// AsWindowsContainerLayerPair indicates that the paths to diff are a pair of
+// Windows Container Layers. The caller must be holding SeBackupPrivilege.
+func AsWindowsContainerLayerPair() WriteDiffOpt {
```

Comment:
I actually dont understand this one? What does this mean to you :)

Reply:
It's like `AsWindowsContainerLayer`, except instead of applying a layer, it's diffing a pair of layers. This whole redirection via `applyFunc`/`writeDiffFunc` is a bit weird, which was my thinking in https://github.com/containerd/containerd/pull/4399#issuecomment-660283335. However, I didn't want to refactor the Apply and Compare paths as part of the same PR that implemented the Compare path to match the Apply path.

Reply:
The place that we actually use this config option, we've already verified the two paths are layers. IMO it makes more sense to make this generic, such as `WithDiffFunc(df diffFunc)` or something like that. What do you think?

Reply:
Hmm. I could refactor `AsWindowsContainerLayer` and `AsWindowsContainerLayerPair` to be `WithApplyFunc` and `WithDiffFunc`, and then `applyWindowsLayer` and `writeDiffWindowsLayer` would move into diff/windows/windows.go, passed in by `(s windowsDiff) Apply` and `(s windowsDiff) Compare` respectively.

However, that means that anyone else trying to use `github.com/containerd/containerd/archive.Apply` or `github.com/containerd/containerd/archive.WriteDiff` would need to provide their own diff funcs to pass in, unless we make `applyWindowsLayer` and `writeDiffWindowsLayer` public, which I'm not super-fond of.

Reply:
Maybe as a clarifying question, how "public" is `github.com/containerd/containerd/archive` support for WCOW layers supposed to be, as an API?

Apart from `github.com/containerd/containerd/diff/windows/`, nothing in this tree seems to be prepared for WCOW usage of `archive.Apply` or `archive.Diff`, so it might make sense to replace `archive.Apply(..., AsWindowsContainerLayer())` and `archive.Diff(..., AsWindowsContainerLayerPair())` with functions directly in `github.com/containerd/containerd/diff/windows/`, and leave `github.com/containerd/containerd/archive` to only support the walking-diff use-case, perhaps with a slight special-case to recognise WCOW layer archives and pretend nothing outside `Files/` exists, i.e. replacing the `strings.Replace(hdr.Name, "Files", "", 1)` in /install.go with one performed during the walk, to allow _some_ rational behaviour for non-Windows platforms manipulating WCOW layer archives.

Reply:
Overall, I'd prefer not to mix such refactorings into this simply because the current approach implements Compare to match Apply, and any further changes should be done to both at once, and as an API change that affects an existing public API, be clearly distinguished as-such.